src/rust/bitbox02-rust/src/hww/api/set_password.rs

```diff
@@ -0,0 +1,45 @@
+// Copyright 2020 Shift Crypto AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::pb;
+use super::{make_error, Error};
+
+use crate::workflow::password;
+use bitbox02::password::Password;
+use core::convert::TryInto;
+use pb::response::Response;
+
+/// Handles the SetPassword api call. This has the user enter a password twice and creates the
+/// seed/keystore. After this call is finished, the keystore is fully unlocked.
+///
+/// `entropy` must be exactly 32 bytes and provides additional entropy used when
+/// creating the seed.
+pub async fn process(pb::SetPasswordRequest { entropy }: &pb::SetPasswordRequest) -> Response {
+    let entropy32: [u8; 32] = match entropy.as_slice().try_into() {
```

Comment:
nanopb had built-in validation of this with the .options file, but nanopb is a bit of an outsider with this. Here and in other protobuf libraries, validation is not part of parsing, so we do it manually.

Reply:
makes sense